src/macros.rs

```diff
@@ -0,0 +1,316 @@
+//! Macro's for easy initialization of date and time values.
+
+/// Create a `NaiveDate` with a statically known value.
+///
+/// Supported formats are 'year-month-day' and 'year-ordinal'.
+///
+/// The input is checked at compile time.
+///
+/// Note: rustfmt wants to add spaces around `-` in this macro.
+/// For nice formatting use `#[rustfmt::skip::macros(date)]`, or use as `date! {2023-09-08}`
+///
+/// # Examples
+/// ```
+/// use chrono::date;
+///
+/// assert_eq!(date!(2023-09-08), date!(2023-251));
+/// ```
+#[macro_export]
+macro_rules! date {
+    ($y:literal-$m:literal-$d:literal) => {{
+        #[allow(clippy::zero_prefixed_literal)]
+        {
+            const DATE: $crate::NaiveDate = match $crate::NaiveDate::from_ymd_opt($y, $m, $d) {
+                Some(d) => d,
+                None => panic!("invalid calendar date"),
```

Comment:
I prefer to repeat back to the user the invalid values.  More specific clues in error messages are always nice.

     panic!("invalid calendar date from values {:?}, {:?}, {:?}", $y, $m, $d)

Reply:
Good comments, thanks @jtmoon79!
I'm not sure if we can use a format string for panics in const context, let me try.

Reply:
Because these macro's are evaluated at compile time the message is already quite nice in my opinion:
```
error[E0080]: evaluation of constant value failed
  --> src\macros.rs:52:20
   |
52 |         assert_eq!(date!(2023-04-31), NaiveDate::from_ymd_opt(2023, 4, 30).unwrap());
   |                    ^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'invalid calendar date', src\macros.rs:52:20
   |
```

Reply:
> Because these macro's are evaluated at compile time the message is already quite nice in my opinion:
> ```
> error[E0080]: evaluation of constant value failed
>   --> src\macros.rs:52:20
>    |
> 52 |         assert_eq!(date!(2023-04-31), NaiveDate::from_ymd_opt(2023, 4, 30).unwrap());
>    |                    ^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'invalid calendar date', src\macros.rs:52:20
>    |
> ```

I still think a nice touch here is to be specific about precisely which argument caused the panic.  Up to you.